### PR TITLE
Add reconstruction graphs & statistics notebooks

### PR DESCRIPTION
**Pull request by rodsan0**

### CODE REVIEW

File: binder/summary-statistics.ipynb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    "stats\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "stats.to_csv('assets/summary_statistics.csv')\n"
```
Comment (mmore500) [inline binder/summary-statistics.ipynb]:
```suggestion
    "stats.to_csv('tree_reconstruction_quality_summary_statistics.csv', index=False)\n"
```

File: binder/summary-statistics.ipynb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    "from teeplot import teeplot\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df = pd.read_csv(\"https://osf.io/ck47r/download\")\n"
```
Comment (mmore500) [inline binder/summary-statistics.ipynb]: add logging details like

```


# prints metadata about notebook runtime
nbm.print_metadata()

df = pd.read_csv("https://osf.io/5trmg/download")
dfdigest = np.bitwise_xor.reduce(
    pd.util.hash_pandas_object(df),
)
print("{:x}".format(dfdigest))
df
```

from https://github.com/mmore500/hstrat-evolutionary-inference/blob/cc6472840ef97e0edacc83fb4d5cf139b22a6c84/binder/perfect-tree-phylometrics.ipynb

File: binder/summary-statistics.ipynb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    "]\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "stats = df.groupby(evolutionary_variables, as_index = False).agg({\n",
```
Comment (mmore500) [inline binder/summary-statistics.ipynb]: can you create a "regime" and "quality" column and then groupby that instead? That way we can just have those columns in the table and not the full details of each parameter

you may need to `.dropna(subset=["regime", "quality"])`

```
df["regime"] = df.apply(
    lambda row: {
        (1, 1, 2, 3.0517578125e-08): "plain",
        (1, 1, 1, 3.0517578125e-08): "neutral selection",
        (1, 1, 8, 3.0517578125e-08): "strong selection",
        (1, 4, 2, 3.0517578125e-06): "weak 4 niche ecology",
        (1, 4, 2, 3.0517578125e-08): "4 niche ecology",
        (1, 8, 2, 3.0517578125e-08): "8 niche ecology",
        (1024, 1, 2, 3.0517578125e-08): "spatial structure",
    }.get(
        tuple(
            row[
                [
                    "num_islands",
                    "num_niches",
                    "tournament_size",
                    "p_niche_invasion",
                ]
            ]
        ),
        np.nan,  # default
    ),
    axis="columns",
)
```

```
df["quality"] = df.apply(
    lambda row: {
        ("reconstructed-tree", 3.0): "33% resolution",
        ("reconstructed-tree", 10.0): "10% resolution",
        ("reconstructed-tree", 30.0): "3% resolution",
        ("reconstructed-tree", 100.0): "1% resolution",
        ("collapsed-phylogeny", 0): "perfect resolution",
    }.get(
        tuple(row[["a", "resolution"]].fillna(0)),
        np.nan,  # default
    ),
    axis="columns",
)
df = df.dropna(axis="index", subset=["quality"]).copy()
df["quality"].unique()
```

also, for ease of use later, I think we should add a column that summarizes all sensitivity analysis variables (i.e., appends them all into a single keyname string). That will make our lives much easier if we want to filter the table, i.e., to only show the main condition in one table and the other sensitivity analysis conditions in another,

```
sensitivity_analysis_variables = [
    "epoch",
    "mut_distn",
]
```

File: binder/summary-statistics.ipynb [hunk 4]
```diff
@@ -0,0 +1,110 @@
+{
```
Comment (mmore500) [inline binder/summary-statistics.ipynb]: name this file reconstruction-quality-summary-statistics.ipynb

File: binder/reconstruction-graphs.ipynb [hunk 5]
```diff
@@ -0,0 +1,121 @@
+{
```
Comment (mmore500) [inline binder/reconstruction-graphs.ipynb]: reconstruction-quality-plots.ipynb

File: binder/reconstruction-graphs.ipynb [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    "from teeplot import teeplot\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df = pd.read_csv(\"https://osf.io/ck47r/download\")\n"
```
Comment (mmore500) [inline binder/reconstruction-graphs.ipynb]: same as below with metadata logging helpers

File: binder/reconstruction-graphs.ipynb [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for group, group_df in df.groupby(methodological_variables + evolutionary_variables):\n",
+    "    teeplot.tee(\n",
```
Comment (mmore500) [inline binder/reconstruction-graphs.ipynb]: you could combine these two in an outer loop over `viz in sns.barplot, pt.RainCloud:`

**Review by mmore500 — COMMENTED** [review]
This is great! A few details to tweak.

File: binder/reconstruction-graphs.ipynb (hunk 6 above)
Comment (mmore500) [reply binder/reconstruction-graphs.ipynb]: *other file not below

File: binder/summary-statistics.ipynb (hunk 1 above)
Comment (rodsan0) [reply binder/summary-statistics.ipynb]: Done in 82e923c

File: binder/summary-statistics.ipynb (hunk 2 above)
Comment (rodsan0) [reply binder/summary-statistics.ipynb]: Done in 589fbbc

File: binder/reconstruction-graphs.ipynb (hunk 6 above)
Comment (rodsan0) [reply binder/reconstruction-graphs.ipynb]: Done in 589fbbc

File: binder/reconstruction-graphs.ipynb (hunk 7 above)
Comment (rodsan0) [reply binder/reconstruction-graphs.ipynb]: Done in 2d28d49

File: binder/summary-statistics.ipynb (hunk 3 above)
Comment (rodsan0) [reply binder/summary-statistics.ipynb]: Done in 3702606

File: binder/summary-statistics.ipynb (hunk 4 above)
Comment (rodsan0) [reply binder/summary-statistics.ipynb]: Renamed in 4d621cc

File: binder/reconstruction-graphs.ipynb (hunk 5 above)
Comment (rodsan0) [reply binder/reconstruction-graphs.ipynb]: Sure, 4d621cc

File: binder/reconstruction-quality-summary-statistics.ipynb [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    "stats\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "stats.to_csv('assets/tree_reconstruction_quality_summary_statistics.csv', index=False)\n"
```
Comment (mmore500) [inline binder/reconstruction-quality-summary-statistics.ipynb:172]: don't save into assets folder; make an outdata folder

File: binder/reconstruction-error-comparisons.ipynb [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    "with pd.option_context(\"display.max_colwidth\", None, \"display.max_rows\", None):\n",
+    "    out_df = (\n",
+    "        pd.DataFrame.from_records(records)\n",
+    "        .sort_values([\"phylometric\", *sensitivity_analysis_variables])\n",
+    "        .reset_index(drop=True)\n",
+    "    )\n",
+    "    display(out_df)\n",
+    "    os.makedirs(\"outdata\", exist_ok=True)\n",
+    "    out_df.to_csv(\n",
+    "        \"a=reconstruction-error-comparisons-between-regimes+ext=.csv\",\n",
```
Comment (mmore500) [inline binder/reconstruction-error-comparisons.ipynb]:
```suggestion
    "        \"outdata/a=reconstruction-error-comparisons-between-regimes+ext=.csv\",\n",
```

File: binder/reconstruction-quality-summary-statistics.ipynb (hunk 8 above)
Comment (mmore500) [inline binder/reconstruction-quality-summary-statistics.ipynb]:
```suggestion
    "os.makedirs(\"outdata\", exist_ok=True)\n",
    "stats.to_csv('outdata/tree_reconstruction_quality_summary_statistics.csv', index=False)\n"
```

**Review by mmore500 — APPROVED** [review]
lgtm, just need to fix the nb build and put data in `outdata/`

File: binder/reconstruction-quality-summary-statistics.ipynb (hunk 8 above)
Comment (rodsan0) [reply binder/reconstruction-quality-summary-statistics.ipynb]: Done, 071607b

**Pull request merged**